pkg/agent/ipassigner/responder/ndp_responder.go

```diff
@@ -138,55 +193,88 @@ func (r *ndpResponder) AddIP(ip net.IP) error {
 	if !utilnet.IsIPv6(ip) {
 		return fmt.Errorf("only IPv6 is supported")
 	}
-	if r.isIPAssigned(ip) {
+
+	target, _ := netip.AddrFromSlice(ip)
+	if r.isIPAssigned(target) {
 		return nil
 	}
-	group, key := parseIPv6SolicitedNodeMulticastAddress(ip)
-	if err := func() error {
-		r.mutex.Lock()
-		defer r.mutex.Unlock()
-		if r.multicastGroups[key] == 0 {
-			if err := r.conn.JoinGroup(group); err != nil {
-				return fmt.Errorf("joining solicited-node multicast group %s for %q failed: %v", group, ip, err)
-			}
-			klog.InfoS("Joined solicited-node multicast group", "group", group, "interface", r.iface.Name)
-		}
-		klog.InfoS("Assigned IP to NDP responder", "ip", ip, "interface", r.iface.Name)
-		r.multicastGroups[key]++
-		r.assignedIPs.Insert(ip.String())
-		return nil
-	}(); err != nil {
+
+	r.mutex.Lock()
+	if err := r.joinMulticastGroup(target); err != nil {
 		return err
 	}
+	r.assignedIPs.Insert(target)
+	r.mutex.Unlock()
+
+	return nil
+}
+
+func (r *ndpResponder) joinMulticastGroup(ip netip.Addr) error {
+	if r.conn == nil {
+		klog.InfoS("NDP responder is not initialized")
+		return nil
+	}
+	group := parseIPv6SolicitedNodeMulticastAddress(ip)
+	if r.multicastGroups[group] > 0 {
+		r.multicastGroups[group]++
+		return nil
+	}
+	if err := r.conn.JoinGroup(group); err != nil {
+		return fmt.Errorf("joining multicast group %s failed: %v", group, err)
+	}
+	klog.InfoS("Joined multicast group", "group", group, "interface", r.linkName)
+	r.multicastGroups[group]++
+	return nil
+}
+
+func (r *ndpResponder) leaveMulticastGroup(ip netip.Addr) error {
+	if r.conn == nil {
+		klog.InfoS("NDP responder is not initialized")
+		return nil
+	}
+	group := parseIPv6SolicitedNodeMulticastAddress(ip)
+	if r.multicastGroups[group] > 1 {
+		r.multicastGroups[group]--
+		return nil
+	}
+	if err := r.conn.LeaveGroup(group); err != nil {
+		return fmt.Errorf("leaving multicast group %s failed: %v", group, err)
+	}
+	klog.InfoS("Left multicast group", "group", group, "interface", r.linkName)
+	delete(r.multicastGroups, group)
 	return nil
 }
 
 func (r *ndpResponder) RemoveIP(ip net.IP) error {
 	if !utilnet.IsIPv6(ip) {
 		return fmt.Errorf("only IPv6 is supported")
 	}
-	r.mutex.Lock()
-	defer r.mutex.Unlock()
-	if !r.assignedIPs.Has(ip.String()) {
+	target, _ := netip.AddrFromSlice(ip)
+	if !r.isIPAssigned(target) {
```

Comment:
I don't really feel good about the fact that we check this separately from the logic below. Usually when you have this pattern:
1) acquire lock
2) check condition
3) release lock
4) acquire lock
5) perform action which depends on above condition
6) release lock

it can be prone to races
it feels like we could hold the lock for the duration of the function

cc @tnqn

Reply:
I agree, I didn't notice the change. This is not a very frequent operation worth extreme performance optimization, we should keep the previous model.

Reply:
reverted the changes of the lock